File: remediation/enrichment/threat_actor_groups.py
"""
Real, MITRE-documented threat-actor-group reference data, correlated to a tenant's own
findings via the ATT&CK technique IDs already tagged on them
(remediation/enrichment/attack_mapping.py).

The 6 groups below are real entries from MITRE ATT&CK's own public Groups catalog
(https://attack.mitre.org/groups/) - id, name, aliases, and associated technique IDs
were independently verified against that live source during implementation (not
extrapolated from memory), same zero-fabrication discipline as this project's
Checkov/Falco rule IDs and NVD CVE data elsewhere.

IMPORTANT - read before citing this anywhere formal: exactly like attack_mapping.py's
own ATT&CK tagging and ai_vuln_taxonomy.py's ATLAS cross-reference, the correlation this
module performs ("this technique already found in your data is associated with this
group") is an **illustrative cross-reference**, not an attribution claim. A technique
being on a group's known-technique list does not mean a given finding was caused by, or
is evidence of, that specific group - many groups share the same common techniques
(phishing, PowerShell, valid-account abuse, etc.). Re-verify any specific group/technique
pairing against attack.mitre.org before citing it in a compliance report or incident
writeup. `aliases`/`summary` are drawn directly from each group's live MITRE page as of
this writing; MITRE updates these pages over time, so re-check before relying on exact
wording.

`status`/`most_recent_activity` (added for the dashboard's "is this group still active"
ask): every group below is independently re-verified (2026-08-05) as currently ACTIVE on
its live MITRE ATT&CK page - none is marked retired/historical, and each page's own most
recently documented campaign is quoted in `most_recent_activity`. This is honestly NOT a
live intrusion-detection or telemetry signal - this app has no live threat-intel feed
ingestion (see the dashboard's own Ingestion disclosure) - it is MITRE's own current
cataloguing status as of the verification date above, the same kind of "real substitute,
not a disguised live feed" honesty this project already applies to its trend charts.

`target_industries` (added for the dashboard's industry-relevance ask): each value is
drawn directly from sector names explicitly named in that group's own live MITRE page (as
of the same 2026-08-05 verification pass), normalized against `INDUSTRIES` below. A group
NOT tagged with a given industry (e.g. no group here carries "Capital Markets" or
"Insurance") genuinely has no sector-specific victimology documented for it on MITRE's own
page today - a real, honest absence (the same "real, not padded" rule this project applies
to every other zero-count taxonomy entry), not an oversight or a placeholder waiting to be
filled in.
"""

# Reference industry/sector taxonomy for the dashboard's group-filter selector. Includes
# every sector named in the user's own request (Financial Services & Banking, Capital
# Markets, Insurance, Healthcare, Retail & Consumer) plus every other sector actually
# evidenced in THREAT_ACTOR_GROUPS' own target_industries below - not every value here is
# guaranteed to match a group (see the module docstring's note on honest absences).
INDUSTRIES = [
    "Financial Services & Banking",
    "Capital Markets",
    "Insurance",
    "Healthcare",
    "Retail & Consumer",
    "Government & Defense",
    "Energy & Utilities",
    "Technology & Telecom",
    "Education",
    "Media & Entertainment",
    "Transportation & Logistics",
    "Hospitality",
]

THREAT_ACTOR_GROUPS = [
    {
        "id": "G0007",
        "name": "APT28",
        "aliases": ["Fancy Bear", "Sofacy", "Sednit", "Pawn Storm", "STRONTIUM", "Forest Blizzard"],
        "mitre_url": "https://attack.mitre.org/groups/G0007/",
        "summary": "Russian state-sponsored group attributed to the GRU (Russia's General "
            "Staff Main Intelligence Directorate), active since at least 2004. Known for "
            "cyber espionage against government, military, and political targets, "
            "including the 2016 DNC/DCCC compromises.",
        "associated_technique_ids": ["T1566", "T1110", "T1003", "T1021", "T1083", "T1059", "T1105", "T1090"],
        "target_industries": ["Government & Defense", "Hospitality"],
        "status": "active",
        "most_recent_activity": "APT28 Nearest Neighbor Campaign (Feb 2022 - Nov 2024)",
    },
    {
        "id": "G0016",
        "name": "APT29",
        "aliases": ["Cozy Bear", "The Dukes", "NOBELIUM", "Midnight Blizzard", "Dark Halo"],
        "mitre_url": "https://attack.mitre.org/groups/G0016/",
        "summary": "Russian state-sponsored group attributed to the SVR (Russia's Foreign "
            "Intelligence Service), active since at least 2008. Targets government "
            "networks across Europe and NATO member countries, research institutions, "
            "and think tanks - known for the 2015 DNC compromise and the 2020-2021 "
            "SolarWinds software supply-chain compromise.",
        "associated_technique_ids": ["T1566", "T1087", "T1059", "T1098", "T1190", "T1078"],
        "target_industries": ["Government & Defense", "Technology & Telecom"],
        "status": "active",
        "most_recent_activity": "Documented activity through at least 2024 (government, cloud/IT supply-chain targeting)",
    },
    {
        "id": "G0032",
        "name": "Lazarus Group",
        "aliases": ["Labyrinth Chollima", "HIDDEN COBRA", "ZINC", "Diamond Sleet"],
        "mitre_url": "https://attack.mitre.org/groups/G0032/",
        "summary": "North Korean state-sponsored group attributed to the Reconnaissance "
            "General Bureau (RGB), active since at least 2009. Responsible for the 2014 "
            "Sony Pictures Entertainment attack; conducts espionage, destructive "
            "attacks, and financially-motivated campaigns. \"Lazarus Group\" is often "
            "used as an umbrella term for multiple North Korean cyber operators.",
        "associated_technique_ids": ["T1566", "T1071", "T1059", "T1087", "T1083", "T1041", "T1547", "T1105"],
        "target_industries": ["Financial Services & Banking", "Media & Entertainment", "Government & Defense", "Energy & Utilities"],
        "status": "active",
        "most_recent_activity": "Operation Dream Job (Sep 2019 - Aug 2020); ongoing per MITRE's most recent page update",
    },
    {
        "id": "G0046",
        "name": "FIN7",
        "aliases": ["Carbon Spider", "ITG14", "Sangria Tempest", "ELBRUS"],
        "mitre_url": "https://attack.mitre.org/groups/G0046/",
        "summary": "Financially-motivated group active since 2013. Initially targeted "
            "retail, restaurant, and financial-services sectors with point-of-sale "
            "malware; since 2020 has shifted toward \"big game hunting\" ransomware "
            "operations (including REvil and its own Darkside RaaS).",
        "associated_technique_ids": ["T1566", "T1059", "T1486", "T1021", "T1087", "T1105", "T1078", "T1547"],
        "target_industries": [
            "Retail & Consumer", "Financial Services & Banking", "Healthcare", "Hospitality",
            "Technology & Telecom", "Media & Entertainment", "Transportation & Logistics", "Energy & Utilities",
        ],
        "status": "active",
        "most_recent_activity": "Documented targeting the United States automotive industry (Apr 2024)",
    },
    {
        "id": "G0034",
        "name": "Sandworm Team",
        "aliases": ["Voodoo Bear", "Telebots", "Seashell Blizzard", "APT44"],
        "mitre_url": "https://attack.mitre.org/groups/G0034/",
        "summary": "Destructive threat group attributed to Russia's GRU military unit "
            "74455, active since at least 2009. Conducted attacks on Ukrainian power "
            "grids (2015, 2016, 2022), the 2017 NotPetya ransomware attack, and "
            "operations against the 2018 Winter Olympics.",
        "associated_technique_ids": ["T1566", "T1105", "T1485", "T1078", "T1570", "T1059", "T1021"],
        "target_industries": ["Energy & Utilities", "Government & Defense", "Transportation & Logistics", "Technology & Telecom"],
        "status": "active",
        "most_recent_activity": "2022 Ukraine Electric Power Attack",
    },
    {
        "id": "G0096",
        "name": "APT41",
        "aliases": ["Wicked Panda", "Brass Typhoon", "BARIUM"],
        "mitre_url": "https://attack.mitre.org/groups/G0096/",
        "summary": "Chinese state-sponsored espionage group that also conducts "
            "financially-motivated operations, active since at least 2012. Targets "
            "healthcare, telecom, technology, finance, education, retail, and video-game "
            "industries across 14+ countries.",
        "associated_technique_ids": ["T1566", "T1190", "T1105", "T1059", "T1087", "T1003", "T1021"],
        "target_industries": [
            "Healthcare", "Technology & Telecom", "Financial Services & Banking",
            "Education", "Retail & Consumer", "Media & Entertainment",
        ],
        "status": "active",
        "most_recent_activity": "APT41 DUST campaign (Jan 2023 - Jun 2024)",
    },
]

_BY_TECHNIQUE_ID = {}
for _group in THREAT_ACTOR_GROUPS:
    for _tid in _group["associated_technique_ids"]:
        _BY_TECHNIQUE_ID.setdefault(_tid, []).append(_group)


def groups_for_technique(technique_id):
    """Returns the (possibly empty) list of THREAT_ACTOR_GROUPS entries associated with
    a given ATT&CK technique ID."""
    return _BY_TECHNIQUE_ID.get(technique_id, [])


def correlate_findings(findings):
    """Returns [{**group, "finding_count": N, "matched_technique_ids": [...]}] for every
    group with at least one associated technique already present in `findings`'
    attack_techniques (see attack_mapping.tag_findings) - sorted by finding_count
    descending. A group absent from the result has none of its known techniques in this
    dataset today, the same "real, not padded" honesty this project applies to every
    other zero-count taxonomy entry elsewhere (rather than always listing all 6 groups
    with a possible zero, callers can decide whether to show the absence explicitly)."""
    technique_ids_seen = set()
    counts_by_technique = {}
    for f in findings:
        for t in f.get("attack_techniques") or []:
            tid = t.get("technique_id")
            if not tid:
                continue
            technique_ids_seen.add(tid)
            counts_by_technique[tid] = counts_by_technique.get(tid, 0) + 1

    results = []
    for group in THREAT_ACTOR_GROUPS:
        matched = [tid for tid in group["associated_technique_ids"] if tid in technique_ids_seen]
        if not matched:
            continue
        finding_count = sum(counts_by_technique[tid] for tid in matched)
        results.append({**group, "finding_count": finding_count, "matched_technique_ids": matched})

    results.sort(key=lambda g: -g["finding_count"])
    return results

File: dashboard/static/js/threatActorGroups.js
// Mirrors remediation/enrichment/threat_actor_groups.py's THREAT_ACTOR_GROUPS on the
// client side, same "small hardcoded reference list mirrored in JS" pattern already
// used by infraTypes.js/scanTypes.js - see that module's docstring for the
// zero-fabrication verification this data went through (each id/name/alias/technique
// list checked directly against attack.mitre.org/groups/ during implementation) and the
// "illustrative cross-reference, not an attribution claim" caveat that applies here too.
//
// `status`/`mostRecentActivity` and `targetIndustries` were added and independently
// re-verified against each group's live MITRE page on 2026-08-05 - see the Python
// module's own docstring for the full sourcing/honesty notes (every group here is
// currently ACTIVE per MITRE, not retired; a group missing an industry tag genuinely has
// no sector-specific victimology documented for it today, not an oversight).
export const INDUSTRIES = [
  "Financial Services & Banking",
  "Capital Markets",
  "Insurance",
  "Healthcare",
  "Retail & Consumer",
  "Government & Defense",
  "Energy & Utilities",
  "Technology & Telecom",
  "Education",
  "Media & Entertainment",
  "Transportation & Logistics",
  "Hospitality",
];

export const THREAT_ACTOR_GROUPS = [
  {
    id: "G0007", name: "APT28",
    aliases: ["Fancy Bear", "Sofacy", "Sednit", "Pawn Storm", "STRONTIUM", "Forest Blizzard"],
    mitreUrl: "https://attack.mitre.org/groups/G0007/",
    summary: "Russian state-sponsored group attributed to the GRU, active since at least 2004 - cyber espionage against government, military, and political targets, including the 2016 DNC/DCCC compromises.",
    associatedTechniqueIds: ["T1566", "T1110", "T1003", "T1021", "T1083", "T1059", "T1105", "T1090"],
    targetIndustries: ["Government & Defense", "Hospitality"],
    status: "active",
    mostRecentActivity: "APT28 Nearest Neighbor Campaign (Feb 2022 - Nov 2024)",
  },
  {
    id: "G0016", name: "APT29",
    aliases: ["Cozy Bear", "The Dukes", "NOBELIUM", "Midnight Blizzard", "Dark Halo"],
    mitreUrl: "https://attack.mitre.org/groups/G0016/",
    summary: "Russian state-sponsored group attributed to the SVR, active since at least 2008 - the 2015 DNC compromise and the 2020-2021 SolarWinds software supply-chain compromise.",
    associatedTechniqueIds: ["T1566", "T1087", "T1059", "T1098", "T1190", "T1078"],
    targetIndustries: ["Government & Defense", "Technology & Telecom"],
    status: "active",
    mostRecentActivity: "Documented activity through at least 2024 (government, cloud/IT supply-chain targeting)",
  },
  {
    id: "G0032", name: "Lazarus Group",
    aliases: ["Labyrinth Chollima", "HIDDEN COBRA", "ZINC", "Diamond Sleet"],
    mitreUrl: "https://attack.mitre.org/groups/G0032/",
    summary: "North Korean state-sponsored group attributed to the Reconnaissance General Bureau, active since at least 2009 - responsible for the 2014 Sony Pictures Entertainment attack.",
    associatedTechniqueIds: ["T1566", "T1071", "T1059", "T1087", "T1083", "T1041", "T1547", "T1105"],
    targetIndustries: ["Financial Services & Banking", "Media & Entertainment", "Government & Defense", "Energy & Utilities"],
    status: "active",
    mostRecentActivity: "Operation Dream Job (Sep 2019 - Aug 2020); ongoing per MITRE's most recent page update",
  },
  {
    id: "G0046", name: "FIN7",
    aliases: ["Carbon Spider", "ITG14", "Sangria Tempest", "ELBRUS"],
    mitreUrl: "https://attack.mitre.org/groups/G0046/",
    summary: "Financially-motivated group active since 2013 - point-of-sale malware against retail/restaurant/financial-services targets, later big-game-hunting ransomware operations.",
    associatedTechniqueIds: ["T1566", "T1059", "T1486", "T1021", "T1087", "T1105", "T1078", "T1547"],
    targetIndustries: [
      "Retail & Consumer", "Financial Services & Banking", "Healthcare", "Hospitality",
      "Technology & Telecom", "Media & Entertainment", "Transportation & Logistics", "Energy & Utilities",
    ],
    status: "active",
    mostRecentActivity: "Documented targeting the United States automotive industry (Apr 2024)",
  },
  {
    id: "G0034", name: "Sandworm Team",
    aliases: ["Voodoo Bear", "Telebots", "Seashell Blizzard", "APT44"],
    mitreUrl: "https://attack.mitre.org/groups/G0034/",
    summary: "Destructive group attributed to Russia's GRU unit 74455, active since at least 2009 - Ukrainian power-grid attacks (2015, 2016, 2022) and the 2017 NotPetya ransomware attack.",
    associatedTechniqueIds: ["T1566", "T1105", "T1485", "T1078", "T1570", "T1059", "T1021"],
    targetIndustries: ["Energy & Utilities", "Government & Defense", "Transportation & Logistics", "Technology & Telecom"],
    status: "active",
    mostRecentActivity: "2022 Ukraine Electric Power Attack",
  },
  {
    id: "G0096", name: "APT41",
    aliases: ["Wicked Panda", "Brass Typhoon", "BARIUM"],
    mitreUrl: "https://attack.mitre.org/groups/G0096/",
    summary: "Chinese state-sponsored espionage group that also conducts financially-motivated operations, active since at least 2012 - targets healthcare, telecom, technology, finance, education, retail, and video-game industries.",
    associatedTechniqueIds: ["T1566", "T1190", "T1105", "T1059", "T1087", "T1003", "T1021"],
    targetIndustries: [
      "Healthcare", "Technology & Telecom", "Financial Services & Banking",
      "Education", "Retail & Consumer", "Media & Entertainment",
    ],
    status: "active",
    mostRecentActivity: "APT41 DUST campaign (Jan 2023 - Jun 2024)",
  },
];

// Returns [{...group, findingCount, matchedTechniqueIds}] for every group with at least
// one associated technique already present in `findings`' attack_techniques - sorted by
// findingCount descending. A group absent from the result simply has none of its known
// techniques in this dataset today (same "real, not padded" honesty as every other
// zero-count taxonomy in this app).
export function correlateFindings(findings) {
  const countsByTechnique = new Map();
  for (const f of findings) {
    for (const t of f.attack_techniques || []) {
      if (!t.technique_id) continue;
      countsByTechnique.set(t.technique_id, (countsByTechnique.get(t.technique_id) || 0) + 1);
    }
  }

  const results = [];
  for (const group of THREAT_ACTOR_GROUPS) {
    const matched = group.associatedTechniqueIds.filter((tid) => countsByTechnique.has(tid));
    if (!matched.length) continue;
    const findingCount = matched.reduce((sum, tid) => sum + countsByTechnique.get(tid), 0);
    results.push({ ...group, findingCount, matchedTechniqueIds: matched });
  }
  results.sort((a, b) => b.findingCount - a.findingCount);
  return results;
}

// Returns the actual finding objects (not just a count) that share at least one of
// `group`'s known ATT&CK techniques - used by the group-detail modal (groupDetail.js)
// to show real assets/owners/findings/remediation-status for one group's overlap with
// this tenant's current data. Accepts either a raw THREAT_ACTOR_GROUPS entry or a
// correlateFindings() result (which already carries the narrower matchedTechniqueIds).
export function findingsForGroup(group, findings) {
  const techniqueIds = new Set(group.matchedTechniqueIds || group.associatedTechniqueIds);
  return findings.filter((f) => (f.attack_techniques || []).some((t) => techniqueIds.has(t.technique_id)));
}
